### ISSUE TITLE
SameDiff performance analysis + prioritization

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#8488]
We will be doing a performance pass on SameDiff during December.

The following tasks need to be completed:
- [ ] Create an op vs. library spreadsheet (what ops can use mkldnn/dnnl, cuDNN, etc)
- [x] ~~Collect op frequency statistics (scan previously collected pretrained models + count op frequency)~~
- [x] ~~Build SameDiff op profiler (output to chrome profiler format)~~
- [x] ~~Build profiler comparison tool (for chrome profiler format)~~
- [x] ~~Run comparisons for SameDiff zoo models we have~~
- [ ] Create prioritized burn-down list of operations to optimize

**Comment by AlexDBlack** [comment]
Op usage statistics for all of the TF frozen models I have:
https://gist.github.com/AlexDBlack/bf7c917e085bcd14a9c2216b4943704e

**Comment by AlexDBlack** [comment]
Profiling results so far:
CPU (avx + mkldnn for both TF 1.15.0 and SD) on linux 3930k (6 core @ 4ghz), 3 warmup + 10 iterations of each. Batch size 32.

Mobilenet V2: https://gist.github.com/AlexDBlack/02daeff0b3d3824be947c8fa59a9963f
Inception Resnet V2: https://gist.github.com/AlexDBlack/05b36f20c967c4c076749885da50484f

Note that the "infinite" ratio means the op was profiled in SD but not TF, in many cases due to TF using fused MKLDNN ops.

Initial issues:
* depthwise conv2d - add MKLDNN/DNNL
* batch norm
* bias add
* relu6/relu
* reshape
* mmul (inception resnet v2, `InceptionResnetV2/AuxLogits/Logits/MatMul`)

**Comment by AlexDBlack** [comment]
```
MobilenetV2/expanded_conv_15/depthwise/depthwise
All arrays are float32, C order and standard strides:
in0: [32, 7, 7, 960]
in1: [3,3,960,1]
out0: [32,7,7,960]
iargs: [3,3,1,1,0,0,1,1,1,1]
```

Bias add, slow case: (3.09 ratio)
```
MobilenetV2/expanded_conv/project/BatchNorm/FusedBatchNorm      (biasadd)
All arrays: float32, C order, default strides
in0: [32,112,112,16]
in1: [16]
out0: [32, 112, 112, 16]
bargs: [false]
```

Bias add, fast case (0.31 ratio)
```
MobilenetV2/expanded_conv_15/project/BatchNorm/FusedBatchNorm   (biasadd)
All arrays: float32, C order, default strides
in0: [32,7,7,160]
in1: [160]
out0: [32,7,7,160]
bargs: [false]
```

```
MobilenetV2/expanded_conv_15/depthwise/BatchNorm/FusedBatchNorm
All arrays: float32, C order, default strides
in0: [32,7,7,960]
in1: [960]
in2: [960]
in3: [960]
in4: [960]
out0: [32,7,7,960]
targs: [0.0010000000474974513]
iargs: [1,1,3]
```

Reshape (9.24 ratio)
```
in0: shape buffer - [2,32,1001,1001,1,8192,1,99]
in1: shape buffer - [1,2,1,131072,1,99] - values (32, 1001)
out0: shape buffer - [2,32,1001,1001,1,8192,1,99]
```

relu6 (ratio 6.05)
```
MobilenetV2/expanded_conv_14/depthwise/Relu6
x = [4,32,7,7,960,47040,6720,960,1,8192,1,99]
z = [4,32,7,7,960,47040,6720,960,1,8192,1,99]
```

relu6 (ratio 0.29)
```
MobilenetV2/expanded_conv_15/expand/Relu6
x = [4,32,7,7,960,47040,6720,960,1,8192,1,99]
z = [4,32,7,7,960,47040,6720,960,1,8192,1,99]
```
Same shape for fast vs. slow relu6 cases :/


```
InceptionResnetV2/AuxLogits/Logits/MatMul
Both inputs float
in0: [2,32,768,768,1,8192,1,99]
in1: [2,768,1001,1001,1,8192,1,99]
out0: [2,32,1001,1001,1,8192,1,99]
iargs: [0,0]
```

**Comment by raver119** [comment]
BiasAdd should be fixed now. Hypothetically.

**Comment by AlexDBlack** [comment]
retesting mobilenet v2 today:
new: https://gist.github.com/AlexDBlack/d668ca009405ae69a2c6f12e9f7f5046
old: https://gist.github.com/AlexDBlack/02daeff0b3d3824be947c8fa59a9963f

Edit: Test was run without AVX2

Faster than TF in all but 2 cases
equal in 1
slower in 1 (see below)

```
MobilenetV2/expanded_conv/project/BatchNorm/FusedBatchNorm
in0: [4,32,112,112,16,200704,1792,16,1,8192,1,99]
in1: [1,16,1,8192,1,99]
out0: [4,32,112,112,16,200704,1792,16,1,8192,1,99]
bargs: [false]
```

**Comment by AlexDBlack** [comment]
AVX2 test results, CPU:

densenet: https://gist.github.com/AlexDBlack/81490a348a987c2cc99e30abf38281a5
squeezenet: https://gist.github.com/AlexDBlack/52dc14ed34735a818397b855b7fb1c4a
nasnet_mobile: https://gist.github.com/AlexDBlack/187ba6ffc660952049df1cb78021afc4
inception_v4: https://gist.github.com/AlexDBlack/21d99add771fb8b11ad3ebdd137835b3
mobilenetv1: https://gist.github.com/AlexDBlack/89759f412e95042276c425a650ea28cf

**Comment by AlexDBlack** [comment]
Also here's AVX2 vs. x86 on inception v4
(note ratio of > 1.0 means AVX2 is slower, <1.0 means AVX2 is faster)

https://gist.github.com/AlexDBlack/05ce0667fb42af8538e4e4d056deb6f1

The identity ops there can be ignored, they aren't actually executed (op is just passed through in java executioner as a no-op)

**Comment by raver119** [comment]
AVX2 is a CPU issue, as discussed in slack.  3rd gen Intel was used for tests.

**Comment by AlexDBlack** [comment]
Retesting bias add, 5930k (no AVX)

(And yes, the op names are ".../FusedBatchNorm", but the actual ops are biasadd)

```
Mobilenetv2:                                                        Before      After       After/Before
MobilenetV2/expanded_conv_1/expand/BatchNorm/FusedBatchNorm         2180973     121334      0.05563
MobilenetV2/expanded_conv_2/expand/BatchNorm/FusedBatchNorm         726714      51455       0.07081
MobilenetV2/expanded_conv/project/BatchNorm/FusedBatchNorm          307602      20557       0.06683
```

**Comment by AlexDBlack** [comment]
Testing master vs. shyrma_depthconv on mobilenetv1: (same linux 5930k as before)
cc @shyrma 

master:
```
 ----- SameDiff Profile Summary -----
Op Name                                                              Op                            Count     Total uS  %         Min       Max       Std       
MobilenetV1/MobilenetV1/Conv2d_1_depthwise/depthwise                 depthwise_conv2d              10        2795995   19.21     253901    292099    14894.03  
MobilenetV1/MobilenetV1/Conv2d_3_depthwise/depthwise                 depthwise_conv2d              10        2648094   18.20     252210    291489    14648.36  
MobilenetV1/MobilenetV1/Conv2d_2_depthwise/depthwise                 depthwise_conv2d              10        1432754   9.85      129738    163916    10330.76  
MobilenetV1/MobilenetV1/Conv2d_5_depthwise/depthwise                 depthwise_conv2d              10        1428315   9.82      129272    161934    11502.95  
MobilenetV1/MobilenetV1/Conv2d_10_depthwise/depthwise                depthwise_conv2d              10        784490    5.39      68698     87920     6245.66   
MobilenetV1/MobilenetV1/Conv2d_9_depthwise/depthwise                 depthwise_conv2d              10        779803    5.36      69920     85529     6664.93   
MobilenetV1/MobilenetV1/Conv2d_11_depthwise/depthwise                depthwise_conv2d              10        778183    5.35      69761     87062     6649.18   
MobilenetV1/MobilenetV1/Conv2d_8_depthwise/depthwise                 depthwise_conv2d              10        747733    5.14      67611     81922     4916.34   
MobilenetV1/MobilenetV1/Conv2d_7_depthwise/depthwise                 depthwise_conv2d              10        717797    4.93      65622     79310     4473.91   
MobilenetV1/MobilenetV1/Conv2d_4_depthwise/depthwise                 depthwise_conv2d              10        622737    4.28      56287     70113     5057.39   
MobilenetV1/MobilenetV1/Conv2d_6_depthwise/depthwise                 depthwise_conv2d              10        315375    2.17      27810     35091     2476.38   
MobilenetV1/MobilenetV1/Conv2d_13_depthwise/depthwise                depthwise_conv2d              10        281535    1.93      21890     33954     3757.71   
MobilenetV1/MobilenetV1/Conv2d_12_depthwise/depthwise                depthwise_conv2d              10        134034    0.92      9808      16005     2210.69

```


shyrma_depthconv:
```
Op Name                                                              Op                            Count     Total uS  %         Min       Max       Std       
MobilenetV1/MobilenetV1/Conv2d_2_depthwise/depthwise                 depthwise_conv2d              10        27051     2.77      2427      3070      218.80    
MobilenetV1/MobilenetV1/Conv2d_3_depthwise/depthwise                 depthwise_conv2d              10        22849     2.34      1914      2970      368.86    
MobilenetV1/MobilenetV1/Conv2d_1_depthwise/depthwise                 depthwise_conv2d              10        22588     2.31      1877      3179      375.39    
MobilenetV1/MobilenetV1/Conv2d_4_depthwise/depthwise                 depthwise_conv2d              10        14077     1.44      1218      1734      183.58    
MobilenetV1/MobilenetV1/Conv2d_5_depthwise/depthwise                 depthwise_conv2d              10        10152     1.04      837       1251      147.27    
MobilenetV1/MobilenetV1/Conv2d_6_depthwise/depthwise                 depthwise_conv2d              10        6870      0.70      552       967       120.71    
MobilenetV1/MobilenetV1/Conv2d_8_depthwise/depthwise                 depthwise_conv2d              10        6016      0.62      477       777       90.03     
MobilenetV1/MobilenetV1/Conv2d_7_depthwise/depthwise                 depthwise_conv2d              10        5877      0.60      480       706       72.82     
MobilenetV1/MobilenetV1/Conv2d_9_depthwise/depthwise                 depthwise_conv2d              10        5750      0.59      472       703       72.98     
MobilenetV1/MobilenetV1/Conv2d_10_depthwise/depthwise                depthwise_conv2d              10        5614      0.57      477       698       73.70     
MobilenetV1/MobilenetV1/Conv2d_11_depthwise/depthwise                depthwise_conv2d              10        5593      0.57      475       888       123.15    
MobilenetV1/MobilenetV1/Conv2d_13_depthwise/depthwise                depthwise_conv2d              10        4524      0.46      395       611       64.10     
MobilenetV1/MobilenetV1/Conv2d_12_depthwise/depthwise                depthwise_conv2d              10        4298      0.44      378       602       63.20     
```


mobilenetv2 is similarly improved.
Still showing slowed than TF, but considerably faster than before MKLDNN support was added

**Comment by agibsonccc** [comment]
Closing — SameDiff has been completely rewritten with a new DSP (Dynamic Shape Plan) execution engine, native plan compilation, CUDA graph replay, and Triton compilation. The architecture referenced in this issue no longer exists. If you still experience this issue with the upcoming release, please open a new issue.
